pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,2039 @@
+        std::vector<std::vector<size_t>> localTgts;
+        localTgts.reserve(tgts.size());
+
+        tgtsVecProcess(this->getNumLocalQubits(), this->getTotalNumQubits(),
+                       tgts, localTgts, tgtsSwapStatus, tgtswirePairs);
+
+        // Check if all target wires are local
+        auto threshold = WiresSwapStatus::Swappable;
+        bool isAllTargetsLocal = std::all_of(
+            tgtsSwapStatus.begin(), tgtsSwapStatus.end(),
+            [&threshold](size_t status) { return status < threshold; });
+
+        mpi_manager_.Barrier();
+
+        if (isAllTargetsLocal) {
+            expvalOnPauliBasis(pauli_words, localTgts, expect_local);
+        } else {
+            size_t wirePairsIdx = 0;
+            for (size_t i = 0; i < pauli_words.size(); i++) {
+                if (tgtsSwapStatus[i] == WiresSwapStatus::UnSwappable) {
+                    auto opsNames = pauliStringToOpNames(pauli_words[i]);
+                    StateVectorCudaMPI<Precision> tmp(
+                        this->getDataBuffer().getDevTag(),
+                        this->getNumGlobalQubits(), this->getNumLocalQubits(),
+                        this->getData());
+
+                    for (size_t opsIdx = 0; opsIdx < tgts[i].size(); opsIdx++) {
+                        std::vector<size_t> wires = {tgts[i][opsIdx]};
+                        tmp.applyOperation({opsNames[opsIdx]},
+                                           {tgts[i][opsIdx]}, {false});
+                    }
+
+                    expect_local[i] =
+                        innerProdC_CUDA(
+                            tmp.getData(), BaseType::getData(),
+                            BaseType::getLength(),
+                            BaseType::getDataBuffer().getDevTag().getDeviceID(),
+                            BaseType::getDataBuffer().getDevTag().getStreamID(),
+                            this->getCublasCaller())
+                            .x;
+                } else {
+                    std::vector<std::string> pauli_words_idx(
+                        1, std::string(pauli_words[i]));
+                    std::vector<std::vector<size_t>> tgts_idx;
+                    tgts_idx.push_back(localTgts[i]);
+                    std::vector<double> expval_local(1);
+
+                    if (tgtsSwapStatus[i] == WiresSwapStatus::Local) {
+                        expvalOnPauliBasis(pauli_words_idx, tgts_idx,
+                                           expval_local);
+                    } else {
+                        applyMPI_Dispatcher(
+                            tgtswirePairs[wirePairsIdx],
+                            &StateVectorCudaMPI::expvalOnPauliBasis,
+                            pauli_words_idx, tgts_idx, expval_local);
+                        wirePairsIdx++;
+                    }
+                    expect_local[i] = expval_local[0];
+                }
+            }
+        }
+
+        auto expect = mpi_manager_.allreduce<double>(expect_local, "sum");
```

Comment:
Do we need to use double for this, or should it be the same as `PrecisionT`?

Reply:
see below!